File: solutions/lua/allergies/2/allergies.lua
local ALLERGIES = {
  "eggs",
  "peanuts",
  "shellfish",
  "strawberries",
  "tomatoes",
  "chocolate",
  "pollen",
  "cats",
}

-- stores the score for each allergy to avoid recomputing
local SCORES = {}
for i, allergy in ipairs(ALLERGIES) do
  SCORES[allergy] = 2 ^ (i - 1)
end

local function list(score)
  local res = {}
  for _, allergy in ipairs(ALLERGIES) do
    if (score & SCORES[allergy]) == SCORES[allergy] then
      table.insert(res, allergy)
    end
  end
  return res
end

local function allergic_to(score, which)
  return (score & SCORES[which]) == SCORES[which]
end

return {
  list = list,
  allergic_to = allergic_to
}

File: solutions/go/allergies/2/allergies.go
package allergies


var allergens = map[string]uint{}

func init() {
    names := []string{"eggs", "peanuts", "shellfish", "strawberries", "tomatoes",
                      "chocolate", "pollen", "cats"}
	for i, name := range names {
        allergens[name] = 1 << i
    }
}

func Allergies(allergies uint) []string {
	res := []string{}
    for name, value := range allergens {
        if allergies & value != 0 {
            res = append(res, name)
        }
    }
    return res
}

func AllergicTo(allergies uint, allergen string) bool {
    return allergies & allergens[allergen] != 0
}
